### PR TITLE
[gh-2233] rename address view and update address view to include nostr.

### LINKED ISSUE #2233: The RPC address argument accepts nostr address format.

### PR DESCRIPTION
**Pull request by feliciss**
## Summary

According to discussion of #2209.

- Closes #2233

### CODE REVIEW

File: crates/rooch-rpc-api/src/jsonrpc_types/address.rs [hunk 1]
```diff
@@ -63,69 +63,88 @@ impl From<RoochAddressView> for AccountAddress {
...
-            Ok(rooch_address) => Ok(StrView(RoochOrBitcoinAddress {
+            Ok(rooch_address) => Ok(StrView(UnitedAddress {
                 rooch_address,
                 bitcoin_address: None,
+                nostr_public_key: None,
             })),
             Err(_) => {
                 let bitcoin_address = BitcoinAddress::from_str(s)?;
-                Ok(StrView(RoochOrBitcoinAddress {
+                let nostr_public_key = NostrPublicKey::from_str(s)?;
```
Comment (jolestar) [inline crates/rooch-rpc-api/src/jsonrpc_types/address.rs:97]: If the str is a nostr address, `BitcoinAddress::from_str(s)?` will raise an error.

**Review by jolestar — COMMENTED** [review]
Need a testcase to use the nostr address.

File: crates/rooch-rpc-api/src/jsonrpc_types/address.rs (hunk 1 above)
Comment (jolestar) [reply crates/rooch-rpc-api/src/jsonrpc_types/address.rs:97]: If it is a Nostr address, it can be convert to Bitcoin address, then to rooch address.

Comment (jolestar) [reply crates/rooch-rpc-api/src/jsonrpc_types/address.rs:97]: The NostrPublicKey's `from_bech32` implemention is not correct.

https://github.com/rooch-network/rooch/blob/be1640ca34c2bd8f071764ffe4614ef9ed398ba6/crates/rooch-types/src/address.rs#L868-L881

NostrPublicKey from_bech32 should return `NostrPublicKey`

Comment (feliciss) [reply crates/rooch-rpc-api/src/jsonrpc_types/address.rs:97]: > The NostrPublicKey's `from_bech32` implemention is not correct.
> 
> https://github.com/rooch-network/rooch/blob/be1640ca34c2bd8f071764ffe4614ef9ed398ba6/crates/rooch-types/src/address.rs#L868-L881
> 
> NostrPublicKey from_bech32 should return `NostrPublicKey`

The `FromBech32`, it is originally for RoochAddress for the type NostrPublicKey to be used for converting to a RoochAddress.

Comment (feliciss) [reply crates/rooch-rpc-api/src/jsonrpc_types/address.rs:97]: There's `impl FromBech32 for XOnlyPublicKey`. The `from_bech32` from NostrPublicKey is for converting to RoochAddress.

Comment (feliciss) [reply crates/rooch-rpc-api/src/jsonrpc_types/address.rs]: `XOnlyPublicKey::from_bech32` could be applied for this scenario.

Comment (feliciss) [reply crates/rooch-rpc-api/src/jsonrpc_types/address.rs]: But if we could convert bech32 string from NostrPublicKey to BitcoinAddress and then to RoochAddress, the NostrPublicKey's from_bech32 implemention could be removed.

Comment (feliciss) [reply crates/rooch-rpc-api/src/jsonrpc_types/address.rs]: > But if we could convert bech32 string from NostrPublicKey to BitcoinAddress and then to RoochAddress, the NostrPublicKey's `from_bech32` implemention could be removed.

We just need to figure out steps for converting from NostrPublicKey to BitcoinAddress and remove the implementation for converting from NostrPublicKey to RoochAddress as BitcoinAddress could convert to RoochAddress.

File: crates/rooch-types/src/address.rs [hunk 2]
```diff
@@ -821,7 +821,20 @@ impl TryFrom<MultiChainAddress> for BitcoinAddress {
...
+
+impl NostrPublicKey {
+    pub fn new(x_only_pk: XOnlyPublicKey) -> Self {
+        Self(x_only_pk)
+    }
+
+    /// Convert from the Nostr XOnlyPublicKey to Bitcoin address
+    pub fn to_bitcoin_address(&self) -> BitcoinAddress {
+        let x_only_pk_bytes = self.0.serialize().to_vec();
+        let bitcoin_address = BitcoinAddress::new(x_only_pk_bytes);
```
Comment (jolestar) [inline crates/rooch-types/src/address.rs:834]: Should not directly use BitcoinAddress::new to convert XOnlyPublicKey to Bitcoin address

File: crates/rooch-rpc-api/src/jsonrpc_types/address.rs (hunk 1 above)
Comment (jolestar) [reply crates/rooch-rpc-api/src/jsonrpc_types/address.rs]: > But if we could convert bech32 string from NostrPublicKey to BitcoinAddress and then to RoochAddress, the NostrPublicKey's `from_bech32` implemention could be removed.

NostrPublicKey also need a from_str, from_str=from_bech32

File: crates/rooch-types/src/address.rs (hunk 2 above)
Comment (jolestar) [reply crates/rooch-types/src/address.rs:834]: `NostrPublicKey` to `PublicKey::Secp256k1` to `BitcoinAddress`

Comment (feliciss) [reply crates/rooch-types/src/address.rs]: > `NostrPublicKey` to `PublicKey::Secp256k1` to `BitcoinAddress`

I referred to BIP-086 to generate a taproot address from `NostrPublicKey `.

File: crates/rooch-rpc-api/src/jsonrpc_types/address.rs (hunk 1 above)
Comment (feliciss) [reply crates/rooch-rpc-api/src/jsonrpc_types/address.rs]: > > But if we could convert bech32 string from NostrPublicKey to BitcoinAddress and then to RoochAddress, the NostrPublicKey's `from_bech32` implemention could be removed.
> 
> NostrPublicKey also need a from_str, from_str=from_bech32

Added.

File: crates/rooch-types/src/address.rs [hunk 3]
```diff
@@ -821,7 +814,32 @@ impl TryFrom<MultiChainAddress> for BitcoinAddress {
...
+        let secp = Secp256k1::verification_only();
+        // new bitcoin taproot address
+        let address = Address::p2tr(
+            &secp,
+            internal_key,
+            None,
+            bitcoin::network::Network::from(network),
+        );
+        // give it to rooch bitcoin struct
+        Ok(BitcoinAddress::new(address.to_string().as_bytes().to_vec()))
```
Comment (jolestar) [inline crates/rooch-types/src/address.rs:840]: BitcoinAddress::from(bitcoin::Address), not address.to_string()

File: crates/rooch-rpc-api/src/jsonrpc_types/address.rs [hunk 4]
```diff
@@ -63,69 +67,101 @@ impl From<RoochAddressView> for AccountAddress {
...
-                Ok(StrView(RoochOrBitcoinAddress {
-                    rooch_address: bitcoin_address.to_rooch_address(),
-                    bitcoin_address: Some(bitcoin_address),
-                }))
-            }
+            Err(_) => match XOnlyPublicKey::from_bech32(s) {
+                Ok(x_only_pk) => {
+                    let nostr_public_key = NostrPublicKey::new(x_only_pk);
+                    let bitcoin_address =
+                        nostr_public_key.to_bitcoin_address(Network::Bitcoin.to_num())?; // TODO: use network variable?
```
Comment (jolestar) [inline crates/rooch-rpc-api/src/jsonrpc_types/address.rs:103]: Rooch BitcoinAddress ignores the network, so using `Network::Bitcoin.to_num()` is ok.

File: crates/rooch-types/src/address.rs (hunk 3 above)
Comment (feliciss) [reply crates/rooch-types/src/address.rs]: Oh Okay.

File: crates/rooch-rpc-api/src/jsonrpc_types/address.rs (hunk 4 above)
Comment (feliciss) [reply crates/rooch-rpc-api/src/jsonrpc_types/address.rs]: Resolved.

**Comment by jolestar** [comment]
Need a test case:

1. Use the nostr address in the CLI argument to ensure the ParsedAddress implemention.
2. Query RPC via the nostr address to ensure the UnitAddress implemention.

**Comment by feliciss** [comment]
> Need a test case:
> 
> 1. Use the nostr address in the CLI argument to ensure the ParsedAddress implemention.
> 2. Query RPC via the nostr address to ensure the UnitAddress implemention.

I have added `account nullify` and `rpc request` for both 1 and 2 test cases.

**Review by jolestar — APPROVED** [review]

**Pull request merged**